text/0000-bee-model-crate/0000-bee-model-crate.md

```diff
@@ -0,0 +1,334 @@
+Besides that, a transaction can only be created from a Transaction Builder, or from a constructor function which expects encoded bytes as received e.g. from a network socket.
+
+```rust
+impl Transaction {
+    
+    pub fn from_bytes(bytes: &Vec<u8>) -> Result<Transaction, TransactionBuilderValidationError> {
+        unimplemented!()
+    }
+
+    pub fn signature_fragments(&self) -> &String {
+        &self.signature_fragments
+    }
+
+    pub fn address(&self) -> &String {
+        &self.address
+    }
+
+    pub fn value(&self) -> &i64 {
+        &self.value
+    }
+
+    pub fn obsolete_tag(&self) -> &String {
+        &self.obsolete_tag
+    }
+
+    pub fn timestamp(&self) -> &i64 {
+        &self.timestamp
+    }
+
+    pub fn current_index(&self) -> &usize {
+        &self.current_index
+    }
+
+    pub fn last_index(&self) -> &usize {
+        &self.last_index
+    }
+
+    pub fn bundle_hash(&self) -> &String {
+        &self.bundle_hash
+    }
+
+    pub fn trunk(&self) -> &String {
+        &self.trunk
+    }
+
+    pub fn branch(&self) -> &String {
+        &self.branch
+    }
+
+    pub fn nonce(&self) -> &String {
+        &self.nonce
+    }
+
+    pub fn tag(&self) -> &String {
+        &self.tag
+    }
+
+    pub fn attachment_timestamp(&self) -> &i64 {
+        &self.attachment_timestamp
+    }
+
+    pub fn attachment_timestamp_lower_bound(&self) -> &i64 {
+        &self.attachment_timestamp_lower_bound
+    }
+
+    pub fn attachment_timestamp_upper_bound(&self) -> &i64 {
+        &self.attachment_timestamp_upper_bound
+    }
+
+}
+```
+
+The Transcation_Builder struct contains public accessible fields. 
+The set values can be validated in the build() function which then returns the constructed transaction.
+Moreover, also the Proof-Of-Work will be called in the build() function.
+
+```rust
+#[derive(Default)]
+pub struct TransactionBuilder {
+    pub signature_fragments: String,
+    pub address: String,
+    pub value: i64,
+    pub obsolete_tag: String,
+    pub current_index: usize,
+    pub last_index: usize,
+    pub bundle_hash: String,
+    pub trunk: String,
+    pub branch: String,
+    pub tag: String,
+    pub attachment_timestamp: i64,
+    pub attachment_timestamp_lower_bound: i64,
+    pub attachment_timestamp_upper_bound: i64,
+}
+
+impl TransactionBuilder {
+
+    pub fn build(self) -> Result<(Transaction, TranscationMetadata), TransactionBuilderValidationError> {
+
+        match TransactionBuilderValidator::validate(&self) {
+            Ok(()) => { ; }
+            Err(e) => { return Err(e) }
+        }
+
+        Ok(Pow::compute(&self))
+
+    }
+
+}
+```
+
+The Pow object takes the Transaction Builder as argument and returns the built transaction together with its metadata.
+Every transaction contains a mutable metadata which consists of following attributes:
+```rust
+pub struct TransactionMetadata {
+    pub transaction_hash: String,
+    pub is_solid: bool,
+    pub snapshot_index: usize
+}
```

Comment:
I think `TransactionMetadata` will be the topic of another RFC so I wouldn't necessarily put the structure here. Or at least not `is_solid` and `snapshot_index`.